src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java

```diff
@@ -33,9 +36,72 @@
     protected String kfsSharedNamespace;
     protected int retryCount;
     
+    private Map<String, String> awsSecretsCache;
+    
+    @Override
+    public void clearCache() {
+        awsSecretsCache = new ConcurrentHashMap<>();
+    }
+    
+    @Override
+    public void logCacheStatus() {
+        if (LOG.isDebugEnabled()) {
+            if (awsSecretsCache == null) {
+                LOG.debug("logCacheStatus, the cache has not been set");
+            } else if (awsSecretsCache.isEmpty()) {
+                LOG.debug("logCacheStatus, the cache has been set, but is empty");
+            } else {
+                for (String key : awsSecretsCache.keySet()) {
+                    LOG.debug("logCacheStatus, AWS secret key: " + key);
+                }
+            }
+        }
+    }
+    
     @Override
     public String getSingleStringValueFromAwsSecret(String awsKeyName, boolean useKfsInstanceNamespace) {
+        createCacheIfNotPresent();
         String fullAwsKey = buildFullAwsKeyName(awsKeyName, useKfsInstanceNamespace);
+        String cachedSecretValue = retrieveSecretFromCache(fullAwsKey);
+        
+        if (StringUtils.isNotBlank(cachedSecretValue)) {
+            return cachedSecretValue;
+        } else {
+            LOG.debug("getSingleStringValueFromAwsSecret, in cache there was no value for " + fullAwsKey);
+            String secretValue = retrieveSecretFromAws(fullAwsKey);
+            updateCacheValue(fullAwsKey, secretValue);
+            return secretValue;
+        }
+    }
+
+    private void createCacheIfNotPresent() {
+        if (awsSecretsCache == null) {
+            LOG.debug("createCacheIfNotPresent, Cache has not been setup, so run the clear cache to create a new cache element");
+            clearCache();
+        }
+    }
+    
+    private String retrieveSecretFromCache(String fullAwsKey) {
+        if (awsSecretsCache.containsKey(fullAwsKey)) {
+            LOG.debug("retrieveSecretFromCache, in cache found a secret value for " + fullAwsKey);
+            return awsSecretsCache.get(fullAwsKey);
+        } else {
+            return null;
+        }
+    }
+    
+    private void updateCacheValue(String fullAwsKey, String secretValue) {
+        if (fullAwsKey == null) {
+            LOG.debug("updateCacheValue, the AWS key provide was null, so can't cache it.");
```

Comment:
This message should say "provided" instead of "provide".

Reply:
ok